Fix to handle hash collisions correctly for DaemonSets

mortent opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**: This adds an integration test for the case where there is a hash collision when creating a ControllerRevision for a DaemonSet. It also fixes a shadowed variable that prevented this functionality from working as intended.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #62519 

**Special notes for your reviewer**:
/sig apps

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Fixes issue when updating a DaemonSet causes a hash collision.
```

mortent committed:
Fix to handle hash collisions correctly for DaemonSet

mortent commented:
CLA should be fixed now

dims commented:
/ok-to-test

mikedanese commented on pkg/controller/daemon/update.go:
```diff
@@ -337,9 +337,9 @@ func (dsc *DaemonSetsController) snapshot(ds *apps.DaemonSet, revision int64) (*
-		done, err := Match(ds, existedHistory)
-		if err != nil {
-			return nil, err
+		done, matchErr := Match(ds, existedHistory)
```
I feel like this is bound to happen again. Can you do something like
```go
if outerErr := err; errors.IsAlreadExists(outerErr) {
  ...
  return outerErr
}
```

mikedanese commented on test/integration/daemonset/daemonset_test.go:
```diff
@@ -740,3 +758,79 @@ func TestInsufficientCapacityNodeWhenScheduleDaemonSetPodsEnabled(t *testing.T)
+		t.Fatalf("Failed to update DaemonSet: %v", err)
+	}
+
+	validateDaemonSetPodsAndMarkReady(podClient, podInformer, 1, t)
```
should you validate that the collision counter is bumped and that the pod has TerminationGracePeriodSeconds set to one?

mortent commented:
Updated to address comments from @mikedanese.

janetkuo commented on test/integration/daemonset/daemonset_test.go:
```diff
@@ -740,3 +782,90 @@ func TestInsufficientCapacityNodeWhenScheduleDaemonSetPodsEnabled(t *testing.T)
+	}
+
+	validateDaemonSetPodsAndMarkReady(podClient, podInformer, 1, t)
+	validateDaemonSetStatus(dsClient, ds.Name, 1, t)
```
`validateDaemonSetPodsAndMarkReady` checks DS has X number of pods and mark them Ready. It doesn't check if the Pods are updated or not. We can skip this step because we don't care if the pods are ready or not. 

`validateDaemonSetStatus` checks whether the status matches or not, but it doesn't check if the Pods are updated or not, either. 

There's a possible race that the Pods haven't been updated before we validate collision count and Pods. Suggest instead wait for new pods and new controllerrevision to be created, and then check DS's collision count.

janetkuo commented on test/integration/daemonset/daemonset_test.go:
```diff
@@ -740,3 +782,90 @@ func TestInsufficientCapacityNodeWhenScheduleDaemonSetPodsEnabled(t *testing.T)
+	one := int64(1)
+	ds.Spec.Template.Spec.TerminationGracePeriodSeconds = &one
+
+	newHash, newName := hashAndNameForDaemonSet(ds)
```
Suggest add comments for each step (or some steps).

janetkuo reviewed (commented):
Thanks for catching the bug!

mortent commented:
@janetkuo Thanks for the detailed review. 
I have updated the validation part of the test to wait for a pod to come up with the updated value for TerminationGracePeriodSeconds and then check that the CollisionCount is the expected value. I think this should avoid the race.
Also added some comments on the key steps in the test code.

janetkuo commented on test/integration/daemonset/daemonset_test.go:
```diff
@@ -740,3 +769,108 @@ func TestInsufficientCapacityNodeWhenScheduleDaemonSetPodsEnabled(t *testing.T)
+		return false, nil
+	})
+	if err != nil {
+		t.Fatal(err)
```
t.Fatalf("Failed to wait for Pods with the latest Spec to be created: %v", err)

janetkuo commented on test/integration/daemonset/daemonset_test.go:
```diff
@@ -740,3 +769,108 @@ func TestInsufficientCapacityNodeWhenScheduleDaemonSetPodsEnabled(t *testing.T)
+	if err != nil {
+		t.Fatalf("Failed to create ControllerRevision: %v", err)
+	}
+
```
nit: save `ds`'s collision count here for comparison later. The count should be count+1 at the end of this test.

mortent commented on test/integration/daemonset/daemonset_test.go:
```diff
@@ -372,6 +374,33 @@ func waitForPodsCreated(podInformer cache.SharedIndexInformer, num int) error {
 }
 
+func waitForDaemonSetCreated(c clientset.Interface, name string, namespace string) error {
+	return wait.Poll(2*time.Second, 30*time.Second, func() (bool, error) {
```
I've updated the test to use PollImmediate and shorter intervals and timeouts. I can do a separate PR to pull out the constants and switch the other tests in the file to use PollImmediate.

mortent commented:
@janetkuo I have updated the PR based on the comments.

mortent commented:
/retest

janetkuo commented on test/integration/daemonset/daemonset_test.go:
```diff
@@ -740,3 +788,112 @@ func TestInsufficientCapacityNodeWhenScheduleDaemonSetPodsEnabled(t *testing.T)
+	var orgCollisionCount int32
+	if ds.Status.CollisionCount != nil {
+		orgCollisionCount = *ds.Status.CollisionCount
+	}
```
Move this down to L846 after DS's CR is created and a fresh DS is got back from the server. This is because collision count might change after CR creation.
